### PR TITLE
2050 ensure windows volume paths are compatible with engine

### PR DESCRIPTION
**Pull request by mnowster**
This fixes absolute windows paths having drives which caused the split on `:` logic to break. It also ensures the path is compatible with engine, which requires a linux path, including the drive.

Fixes https://github.com/docker/compose/issues/2050

### CODE REVIEW

File: compose/service.py [hunk 1]
```diff
@@ -936,20 +937,51 @@ def build_volume_binding(volume_spec):
...
     parts = volume_config.split(':')
+
+    if IS_WINDOWS_PLATFORM:
+        # relative paths in windows expand to include the drive, eg C:\
+        # so we join the first 2 parts back together to count as one
+        drive, volume_path = os.path.splitdrive(volume_config)
+        windows_parts = volume_path.split(":")
+
+        if drive:
+            windows_parts[0] = "/{}{}".format(drive.replace(':', '/'), windows_parts[0])
```
Comment (aanand) [inline compose/service.py]: Reading `ensure_engine_compatibility`'s docstring, I sort of expected this bit of logic to happen over there. Would it be simple enough to move it there?

Comment (mnowster) [inline compose/service.py]: Yep, good point. I'll move it.

File: compose/service.py [hunk 2]
```diff
@@ -936,20 +937,55 @@ def build_volume_binding(volume_spec):
     return volume_spec.internal, "{}:{}:{}".format(*volume_spec)
 
 
+def ensure_engine_compatibility(external_path, internal_path):
```
Comment (dnephin) [inline compose/service.py]: Minor naming nit: this function name seems a lot more general then what it's doing. "ensure" to me means "validate", and "engine compatibility" is a much larger issue than just paths.

How about `normalize_host_path()`  (or something along those lines) ?

File: compose/service.py [hunk 3]
```diff
@@ -936,20 +937,55 @@ def build_volume_binding(volume_spec):
...
     else:
-        external = os.path.normpath(parts[0])
-        internal = os.path.normpath(parts[1])
+        external, internal = ensure_engine_compatibility(os.path.normpath(parts[0]), os.path.normpath(parts[1]))
 
-    mode = parts[2] if len(parts) == 3 else 'rw'
+    try:
+        mode = parts[2]
+    except IndexError:
+        mode = 'rw'
```
Comment (dnephin) [inline compose/service.py]: Why this change from if/else to try/except? If we expected the common case to be that 'rw' was specified, than I think this would be faster, but I think the more common case is actually no mode, which makes the if/else much faster.

File: compose/service.py (hunk 2 above)
Comment (mnowster) [inline compose/service.py]: This was a tough one to name, I was nearly going to invent a word :crystal_ball: but I chatted with @aanand about it as well. I originally was going to use `transform_for_engine` or something similar so that it is very explicit that we are changing the way these paths look for the specific need of the engine.
`normalize_host_path` doesn't capture the specific need of we are doing this for the engine. I understand the point that there are larger issues than paths, could be `ensure_paths_are_engine_compatible` is clearer?

File: compose/service.py (hunk 3 above)
Comment (mnowster) [inline compose/service.py]: Aah ok, I thought the common case was it would be specified, if it's the other way around then yeh, agree, I'll change it back to an if.

File: compose/service.py (hunk 2 above)
Comment (dnephin) [inline compose/service.py]: Does it matter why we're doing it? I usually keep naming to "what" it's doing, and leave the "why" for the docstring.

To me this is a "normalize" operation not "ensure". "ensure" is validation/assertions, this is "do something in some cases, but ignore it in others".

`normalize_paths_for_engine()` would work for me (even though "engine" might not always be relevant, it's more that we're normalizing them for posix, so my first choice would probably be `normalize_paths_to_posix()` or something like that).

Comment (mnowster) [inline compose/service.py]: That naming, was me trying to describe "what" we're doing and part of that what included the "why". When naming things, especially tricky things like this, I try to think about the person who has to come after me and fix it or modify the code, so I lean towards verbosity and whys as that will aid them to understand why the code is the way it is and pick it up quicker than it took me to figure it out the first time.

Ultimately we're agreeing with each other but with different words :)

I'm happier with `normalize_paths_for_engine` for now, as this is a workaround a hack within the engine itself. Then when that is no longer the case, we can rename it again. Cool.

**Comment by aanand** [comment]
LGTM
